projects/vdk-core/tests/functional/run/test_run_load_setuptools_errors.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import unittest
+from unittest.mock import MagicMock
+from unittest.mock import patch
+
+from vdk.internal.cli_entry import main
+
+
+class TestImportErrors(unittest.TestCase):
+    @patch(
+        "pluggy._manager.PluginManager.load_setuptools_entrypoints",
+        side_effect=ImportError("Test import error."),
+    )
+    @patch("os._exit", side_effect=Exception("VDK run termination exception."))
+    @patch(
+        "vdk.internal.builtin_plugins.termination_message.file_util.WriteToFileAction.write_to_file"
+    )
+    def test_import_error(
+        self,
+        file_util: MagicMock,
+        patched_exit: MagicMock,
+        load_setuptools_entrypoints: MagicMock,
+    ):
+        """
+        This test simulates a scenario where a user imports a library that
+        has a clash with vdk internal imports. The desired behaviour is to
```

Comment:
> This test simulates a scenario where a user imports a library that
> has a clash with vdk internal imports.

By reading the description, my impression about the use-case is - the user code performs a non-valid import that clashes with vdk internal imports. Then, why patching pluggy's PluginManager? Could you provide a real-world stack trace example?
Adding a functional test will make the use-case more clear

Reply:
Here is a sample stack trace of the problem:

```
Traceback (most recent call last):
  File "/vdk/site-packages/vdk/internal/plugin/plugin.py", line 52, in load_plugins_from_setuptools_entrypoints
    self.__plugin_manager.load_setuptools_entrypoints(self.__group_name)
  File "/vdk/site-packages/pluggy/manager.py", line 299, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/vdk/site-packages/importlib_metadata/__init__.py", line 203, in load
    module = import_module(match.group('module'))
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/vdk/site-packages/supercollider/vdk/supercollider_vdk_plugin.py", line 10, in <module>
    from supercollider.vdk.api.load_plugin import LoadPlugin
  File "/vdk/site-packages/supercollider/vdk/api/load_plugin.py", line 6, in <module>
    from supercollider.vdk.api.load import ImpalaLoader
  File "/vdk/site-packages/supercollider/vdk/api/load.py", line 15, in <module>
    import pandas
  File "/vdk/site-packages/pandas/__init__.py", line 22, in <module>
    from pandas.compat import (
  File "/vdk/site-packages/pandas/compat/__init__.py", line 15, in <module>
    from pandas.compat.numpy import (
  File "/vdk/site-packages/pandas/compat/numpy/__init__.py", line 27, in <module>
    f"this version of pandas is incompatible with numpy < {_min_numpy_ver}\n"
ImportError: this version of pandas is incompatible with numpy < 1.17.3
your numpy version is 1.16.5.
Please upgrade numpy to >= 1.17.3 to use this pandas versionThe above exception was the direct cause of the following exception:Traceback (most recent call last):
  File "/vdk/vdk", line 8, in <module>
    sys.exit(main())
  File "/vdk/site-packages/vdk/internal/cli_entry.py", line 165, in main
    plugin_registry.load_plugins_from_setuptools_entrypoints()
  File "/vdk/site-packages/vdk/internal/plugin/plugin.py", line 67, in load_plugins_from_setuptools_entrypoints
    raise PluginException(message) from e
vdk.api.plugin.plugin_registry.PluginException: Plugin load failed
  WHAT HAPPENED : Cannot load plugin from setuptools entrypoint for group vdk.plugin.run
WHY IT HAPPENED : See exception for possible reason
   CONSEQUENCES : The CLI tool will likely abort.
COUNTERMEASURES : Re-try again. Check exception mesage and possibly uninstall a bad plugin (pip uninstall) Or see what plugins are installed (use `pip list` command) and if there are not issues. Or try to reinstall the app in a new clean environment.Try to revert to previous version of hte CLI tool.If nothing works open a SRE ticket 
```
The import itself is not invalid, it is incompatible with internal imports. The problem here is that the error occurs in the cli_entry.py - main() method which is unfortunately not covered by the CliEntryBasedTestRunner() used in our functional tests another problem is that we need that specific version of numpy and pandas to reproduce it, meaning that we have to install those dependencies in the test environment, that's why I believe it's more convinient to mock the specified method to throw the same type of exception instead.

Reply:
Stacktrace shows the issue occurs during `plugin_registry.load_plugins_from_setuptools_entrypoints()` when loading plugins, before any data job user code had been handled yet. 

At the same time:

> why: some user data jobs have been failing with ImportErrors (where a
> user import clashes with internal imports, however this crashed the
> vdk runtime without writing a termination message.

How come the plugin load lifecycle phase - clashes with user code?

Reply:
I don't know the intricacies of it all, but it crashes when the user has imported a library that is incompatible with one of our internal ones. So I suppose it fails when the interpreter goes over an import statement in a plugin that clashes with a user import statement.

Reply:
I see, https://github.com/vmware/versatile-data-kit/blob/c2444dfe82c4ffeff378988d2e7dc26a850fbdd9/projects/control-service/projects/job-builder-rootless/Dockerfile.python.vdk#L26 -> data job dependencies come preinstalled on image build, that is triggered on `vdk deploy`.  

If you want to deep-dive, then what are the pros and cons of dependency compatibility validation during image build vs run phase?
On another note, what job builder are you using, standard/rootless/secure?

Reply:
I think we discussed this with Toni, and concluded that the fix should be in vdk-core, since a job builder is decoupled from vdk and this won't entirely resolve the issue for local runs or already deployed jobs with this issue.

Reply:
In that case the customer would eventually understand the dependency issues not early on deployment, but late when the first time data job I scheduled and runs. Context switching UX is a downside to me, a user friendly workflow includes both checks. However I think this may be out of scope of this PR.